src/throughput.rs

```diff
@@ -0,0 +1,115 @@
+use std::fmt::Debug;
+use std::fmt::{self, Display};
+use std::time::Duration;
+use std::time::Instant;
+
+/// Entry for throughput monitor with measured information.
+struct ThroughputEntry {
+    measured_throughput: u32,
+    start: Instant,
+}
+
+impl ThroughputEntry {
+    /// Construct a new throughput entry.
+    pub fn new(measured_throughput: u32, time: Instant) -> ThroughputEntry {
+        ThroughputEntry {
+            measured_throughput,
+            start: time,
+        }
+    }
+}
+
+/// Helper to monitor throughput.
+///
+/// Throughput is calculated at some duration.
+/// For each duration an entry is created to keep track of the history of throughput.
+///
+/// With this type you can calculate the average or get the total and last measured throughput.
+pub struct ThroughputMonitoring {
+    throughput_duration: Duration,
+    timer: Instant,
+    current_throughput: u32,
+    measured_throughput: Vec<ThroughputEntry>,
+}
+
+impl ThroughputMonitoring {
+    /// Construct a new instance of `ThroughputMonitoring`
+    pub fn new(throughput_duration: Duration) -> ThroughputMonitoring {
+        ThroughputMonitoring {
+            throughput_duration,
+            timer: Instant::now(),
+            current_throughput: 0,
+            measured_throughput: Vec::new(),
+        }
+    }
+
+    /// This will increase the throughput by one, when the `throughput_duration` has elapsed since the last call, then an throughput entry will be created.
+    pub fn tick(&mut self) -> bool {
+        if self.timer.elapsed() >= self.throughput_duration {
+            self.measured_throughput
+                .push(ThroughputEntry::new(self.current_throughput, self.timer));
+            self.current_throughput = 0;
+            self.timer = Instant::now();
+            return true;
+        } else {
+            self.current_throughput += 1;
+            return false;
+        }
+    }
+
+    /// Returns the average throughput over all throughput up-till now.
+    pub fn average(&self) -> u32 {
+        if self.measured_throughput.len() != 0 {
```

Comment:
Instead of storing all of the intermediate measurements, it may make sense to calculate a running average rather than a total average. Would lessen the need to store all `ThroughputEntry`s.

Reply:
Agree and disagree, agree because we aren't using the most out of this storage, disagree because we also store the time, of the measurement, this time will be useful to see throughput at certain times.